### ISSUE TITLE
[Feature request] facilitated import of JoinMarket wallets

### ISSUE THREAD

**Issue by openoms** [cryptoadvance/specter-desktop#990]
Background:
Specter Desktop is ideal to manage multiple watch only wallets and construct transactions ready for signing. The `Wallets overview` is especially helpful to track the global balance and transaction activity.

Since https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/631352c4531da01c5c58ef8bd347c65fb49c93aa it is possible to sign extrenally prepared PSBTS-s: https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/docs/USAGE.md#co-signing-a-psbt

Issue:
In relation to the JoininBox project https://github.com/openoms/joininbox/pull/26 I am trying to find the best way to quickly import a JM wallet (with it's 5 mixdepth/accounts) to Specter.
The manual process works (especially since JM defaults to native SegWit wallets), but there could be an easier option.
We could either generate an info file to be imported to Specter or there could be an option to simply paste the 5 xpubs to create the JM device and wallets.
Similar to Specter JoinMarket uses a watch only Bitcoin Core wallet to track the chain, so even that could be helpful. Possibly would not even need a rescan if would be able to use the existing wallet? Even the (automatically applied) labels could be kept in that case.
What do you think the best approach would be?

For an example here are the public keys of a sample wallet on signet which I am using for testing: https://termbin.com/yd2k
Xpubs only:
```
$ python wallet-tool.py signet|grep mixdepth|sed -n '1~2p'|awk '{print $1,$2,$3}'
Enter passphrase to decrypt wallet: 
mixdepth 0 tpubDCqJ97RNYRNeVxzyXHWzYfehVqiJaCDwPRmcFp8a2ScEN8WXU1KzvRgg85CxHT8tq8CYijTubKZtJos3USjUKbwZDWxv3JuySFHUhqBafeB
mixdepth 1 tpubDCqJ97RNYRNeYCQj82eDFmpadSqLZNrJjLXTqUPncEUrnWwP1jzEXtA6zV3CU4h5Q8ZBBXBJdxSGxMqpMhQEQ2rzhXgHcb9Smh8AUQqtwUE
mixdepth 2 tpubDCqJ97RNYRNebUfUFBuLetjqcNguze1S9N9C8cuwViL6ur6Q3Lq3UvBtJvCUB4rkqpFDj7VTpHaYshLj6nkpqoApzXX3k6kYvhDnv2E5u5z
mixdepth 3 tpubDCqJ97RNYRNecxHXqaszJaWYQkfGDZdAMdoXeS8AkaM6LV4c2hr3fUefndmW9wzgA7J8wbAQYjZvxfZqaxro3Sfvn5WtZMLurxGUrXKLgTE
mixdepth 4 tpubDCqJ97RNYRNegb2p3GMMmSHjz5xuNKg63HsnoQKchFCkKADrHKk5ox59hLYJtWY6hMeRCQmHiiTpP3ff4W1W6xjzLCYEERbsYPw8EhzzXaW
```

**Comment by openoms** [comment]
After exploring further the existing method is quite user friendly so will share some screenshots here:
Click `Add new device`:
![image](https://user-images.githubusercontent.com/43343391/109691056-f0bdd300-7b7e-11eb-9d62-f60342cb60e8.png)
then `Manual configuration`:
![image](https://user-images.githubusercontent.com/43343391/109691183-1814a000-7b7f-11eb-942a-932a0fdd74f0.png)
Source the master public keys:
Name the wallet, select the `Other` device type, paste the xpubs and `Continue`:
![image](https://user-images.githubusercontent.com/43343391/109691535-7477bf80-7b7f-11eb-8fc3-c68f5d464c6a.png)
Looking good:
![image](https://user-images.githubusercontent.com/43343391/109691656-92ddbb00-7b7f-11eb-8d86-fe9d652b5068.png)
Now `Add new wallet`, `Single key wallet`, pick the previously added device.
Will need to do this one-by-one for the 5 mixdepths:
![image](https://user-images.githubusercontent.com/43343391/109691962-e0f2be80-7b7f-11eb-96cf-ef10d3c7384a.png)
With SigNet the balance and transactions show up almost immediately:
![image](https://user-images.githubusercontent.com/43343391/109692137-113a5d00-7b80-11eb-9a8e-99c2ddad31ab.png)
Once finished the 5 mixdepths show up in the `Wallets overview`:
![image](https://user-images.githubusercontent.com/43343391/109692506-860d9700-7b80-11eb-99fb-e7c44ce3bdb6.png)

In summary with the manual configuration it is user friendly already. 
Maybe a specific Device type could be introduced for JoinMarket to have the 5 wallets created automatically without going through the manual process 5 times.
Other question if it would be possible to use the watch-only Bitcoin Core to have the labels synchronised?

**Comment by ChristopherA** [comment]
BTW, you might want to support add additional requirement for using the latest QR/UR research on `crypto-request` but for joinmarket. see: https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2021-001-request.md

Some rough text from a teaser I'm working on:
---

Before we can create a multisig, we need all the public keys. If the network app is creating the account policy, it also easily creates its own first xpub `xpub-A`.

Now it needs a second xpub `xpub-B` from each of the offline devices. The network wallet displays this `cypto-request` QR code:

Which decodes to:

![](https://i.imgur.com/WaZgHJU.png =400x400)

```
UR:CRYPTO-REQUEST/OEADTPDAGDOELBEOLPBEPSGTOERDDKPKBYAHWNSNDYAOTAADYKOTADYKAOTAADDYOEADLOCSDYYKADYKAEYKAOYKAXAAAXTAADEHOYAOADTTFRJYSF
```

This QR UR asks for the offline wallet to give an xpub from any seed under its control, for the `[48’/1’/0’/2’]` derivation (in this case: multisig, bitcoin, account 0, segwit). 

The offline wallet then asks the holder to select a seed to derive the key from, or possibley to create a new seed dedicated to this multisig. Some offline wallets can only support one seed, so it will just ask the holder's consent to share the xpub. It returns this QR:


![](https://i.imgur.com/wfCYYuC.png)

ur:crypto-response/oeadtpdagdmhzmhdlyaoaeaeaepftdgljladaeaeaeaotaaddlonaxhdclaxtebkfpzmpminjpjngdinrsiatphplkwzmofrspztlrpegeplihonvaenrkhdhseoaahdcxuedybyvytpcscljsgagmpejecsgoenoxghecbdwmbdktfwhfkoutlfvlkslygakgahtaadehoyaoadamtaaddyotadlocsdyykadykaeykaoykaocyhngrmuwzaxaaaycyjtieadltdiutcywf


----

![](https://i.imgur.com/cioPsod.png =400x400)

Which decodes to:

```
UR:CRYPTO-HDKEY/ONAXHDCLAXTEBKFPZMPMINJPJNGDINRSIATPHPLKWZMOFRSPZTLRPEGEPLIHONVAENRKHDHSEOAAHDCXUEDYBYVYTPCSCLJSGAGMPEJECSGOENOXGHECBDWMBDKTFWHFKOUTLFVLKSLYGAKGAHTAADEHOYAOADAMTAADDYOTADLOCSDYYKADYKAEYKAOYKAOCYHNGRMUWZAXAAAYCYJTIEADLTLEZSVSLB
```

Decoding this in CBOR you get a `crypt-hdkey`:

```
[604b93f2/48'/1'/0'/2']xpub6ELxvfv7kLAjWvZmfP9rkH4dFrsRJn2JeWfzyzwUsHHTcPKg17ELYMtAu2T8f9H8HQfDWv3g4dsk7yrMY5VJuhreB4JXaTHsZyegqJS1FM6
```

> or in the SLIP-132 form (inproperly IMHO)
>
> `tpubDEibSujoTc8Bnikd7a8wxCPzayy5JAXNC9GJLYzdDC3MMfzP5L5RgoZ194XvBeEN5KC88VZYbjhnpALxvwLYo9JwEf3qFyx79wF6bKnukNM`

**Comment by openoms** [comment]
@ChristopherA thank you, will keep the QRcode format in mind when communicating the publix keys between devices. 
Currently my main usecase for the JM wallet in Specter is to open lightning channels with the PSBTs constructed in the Specter GUI. Would not use JM in a multisig setup since it is meant to be a hot wallet and there are funds on mulitple (5) singel sig derivation paths already.

Some added info on the labels in JM from @AdamISZ:
In the watch only Bitcoin Core wallet the JM addresses the "addresses are labelled `joinmarket-wallet-XXXXX` where XXXXX is basically some intended-to-be-unique truncated hash of the wallet (it goes from the first address)" : https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/365ca38cc2a821039c501b31be264b0ecffaefef/jmclient/jmclient/wallet_utils.py#L313
Then the labels like: 
`cj-out`, `change-out`, `non-cj-change`, `deposit` are applied automatically in the JM wallet and not even stored in the wallet file currently:
 https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/wallet_utils.py#L412

Specter could do a basic transaction analysis like this to apply labels automatically, but it might be out of scope.

Regarding using the already present watch-only Bitcoin Core wallet I think Specter could filter and import the addresses if given the labels.
Since the address labels in Core are generated from the first addres it  only need access to the watch-only Core wallet file and the first address.

**Comment by kristapsk** [comment]
> Then the labels in the JoinMarket wallet like:
> `cj-out`, `change-out`, `non-cj-change`, `deposit` are applied automatically

I would not call them "labels", they are more like statuses, indicating their relationship with incoming transaction(s). Manual address labeling support is something I plan to work on to add to JM in future (mostly to label external receiving addresses I have shared to third parties, to avoid accidental address reuse). But will be stored in JM wallet.jmdat files only likely.
